internal/serialization/internal_serialization_methods_it_test.go

```diff
@@ -0,0 +1,198 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package serialization_test
+
+import (
+	"context"
+	"reflect"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"github.com/hazelcast/hazelcast-go-client"
+	"github.com/hazelcast/hazelcast-go-client/internal/it"
+	iserialization "github.com/hazelcast/hazelcast-go-client/internal/serialization"
+	"github.com/hazelcast/hazelcast-go-client/serialization"
+)
+
+func TestInternalSerializationMethods(t *testing.T) {
+	testCases := []struct {
+		name string
+		f    func(t *testing.T)
+	}{
+		{name: "DefaultCompactSerializer", f: defaultCompactSerializerTest},
+		{name: "DefaultPortableSerializer", f: defaultPortableSerializerTest},
+		{name: "OverrideBuiltinSerializer", f: overrideBuiltinSerializerTest},
+	}
+	for _, tc := range testCases {
+		t.Run(tc.name, tc.f)
+	}
+}
+
+func defaultCompactSerializerTest(t *testing.T) {
+	var sc serialization.Config
+	sc.Compact.SetSerializers(FooCompactSerializer{})
+	tcx := it.MapTestContext{
+		T: t,
+		ConfigCallback: func(tcx it.MapTestContext) {
+			tcx.Config.Serialization = sc
+		},
+	}
+	tcx.Tester(func(tcx it.MapTestContext) {
+		ctx := context.Background()
+		hazelcast.SetDefaultCompactDeserializer(DefaultCompactDeserializer{})
+		require.NoError(t, tcx.M.Set(ctx, "foo", NewFoo("abc")))
+		cfg2 := *tcx.Config
+		cfg2.Serialization.Compact = serialization.CompactConfig{}
+		require.NoError(t, cfg2.Validate())
+		client2 := it.MustClient(hazelcast.StartNewClientWithConfig(ctx, cfg2))
+		defer client2.Shutdown(ctx)
+		m2 := it.MustValue(client2.GetMap(ctx, tcx.MapName)).(*hazelcast.Map)
+		v := it.MustValue(m2.Get(ctx, "foo"))
+		require.Equal(t, NewFoo("OK"), v)
+	})
+}
+
+func defaultPortableSerializerTest(t *testing.T) {
+	hazelcast.SetDefaultPortableDeserializer(DefaultPortableDeserializer{})
+	var cfg1 serialization.Config
+	cfg1.SetPortableFactories(&FooPortableFactory{})
+	ss1 := mustSerializationService(iserialization.NewService(&cfg1, nil))
+	data := mustData(ss1.ToData(&FooPortableSerializer{foo: NewFoo("abc")}))
+	var cfg2 serialization.Config
+	ss2 := mustSerializationService(iserialization.NewService(&cfg2, nil))
+	v := it.MustValue(ss2.ToObject(data))
+	assert.Equal(t, &FooPortableSerializer{foo: NewFoo("OK")}, v)
+}
+
+func overrideBuiltinSerializerTest(t *testing.T) {
+	backup := iserialization.BuiltinDeserializers[iserialization.TypeInt32]
+	hazelcast.SetBuiltinDeserializer(Int32Deserializer{})
+	defer func() {
+		iserialization.BuiltinDeserializers[iserialization.TypeInt32] = backup
+	}()
+	var cfg serialization.Config
+	ss := mustSerializationService(iserialization.NewService(&cfg, nil))
+	data := mustData(ss.ToData(int32(100)))
+	v := it.MustValue(ss.ToObject(data))
+	assert.Equal(t, int32(38), v)
+}
```

Comment:
CLC uses the client, but can also access the client's serialization service.

Reply:
what about testing both in all tests then?

Reply:
We have to test the first one with a client, since in order to trigger schema loading, serialization service shouldn't have it. But a connection to the cluster is required to fetch the schema, so serialization service on its own is not sufficient.

For portable and builtin serializer tests, connection to the cluster is not required, so using a client is not necessary.